delegate-app/HydraAuction/Delegate/Server.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE RankNTypes #-}
+
+module HydraAuction.Delegate.Server (
+  -- * Delegate server types
+  DelegateServerConfig (..),
+  DelegateServerLog (..),
+  DelegateError (..),
+
+  -- * tracing types
+
+  -- ** tracing transformer
+  TracerT (..),
+  runWithTracer,
+  runWithTracer',
+  mapTracerT,
+  askTracer,
+
+  -- ** tracing type class
+  MonadTracer (..),
+
+  -- ** delegate tracing
+  DelegateTracer,
+
+  -- * wai extras
+  ServerAppT,
+) where
+
+-- Prelude imports
+import Hydra.Prelude (ReaderT (runReaderT), withReaderT)
+import Prelude
+
+-- Haskell imports
+import Control.Tracer (Contravariant (contramap), Tracer, natTracer, traceWith)
+import GHC.Generics (Generic)
+import Prettyprinter (Pretty (pretty), line, viaShow)
+
+-- Cardano imports
+import Hydra.Network (IP, PortNumber)
+
+-- Hydra auction imports
+
+import Control.Monad.RWS
+import HydraAuction.Delegate.Interface (DelegateResponse, FrontendRequest)
+import Network.WebSockets (PendingConnection)
+
+-- | The config for the delegate server
+data DelegateServerConfig = DelegateServerConfig
+  { dlgt'host :: IP
+  -- ^ the host of the delegate server
+  , dlgt'port :: PortNumber
+  -- ^ the port number the delegate server receives input at
+  , dlgt'tick :: Int
+  -- ^ the amount of milliseconds, the thread should wait
+  }
+
+data DelegateServerLog
+  = Started PortNumber
+  | FrontendConnected
+  | FrontendInput FrontendRequest
+  | DelegateOutput DelegateResponse
+  | DelegateError DelegateError
+  deriving stock (Eq, Show, Generic)
+
+instance Pretty DelegateServerLog where
+  pretty = \case
+    Started port -> "Started Server at Port" <> pretty (show port)
+    FrontendConnected -> "Frontend connected to Server"
+    DelegateOutput out -> "Delegate output" <> line <> viaShow out
+    FrontendInput inp -> "Frontend input" <> line <> viaShow inp
+    DelegateError err -> "Delegate error" <> line <> pretty err
+
+-- FIXME: needs actual error
+data DelegateError = FrontendNoParse
```

Comment:
What would we do with error datatype?

Client should get error by async response anyway. So I guess this should be fit in regular `delegateStep` logic.

Reply:
well those are the errors that are emitted by the thing that calls `execDelegateRunner`, so we cannot emit them from inside the monad, e.g. this specific appears if we cannot parse the JSON that the frontend inputs

Reply:
Well yeah, this doesn't get enqueued anywhere

Reply:
Reminder to self to respond with this error if there occurs a failure in encoding